introduction/related-work.tex

```diff
@@ -0,0 +1,95 @@
+
+\subsection{Abstraction Level}
+In most computer systems, the hardware interaction and resource management is abstracted away with an \gls{os}.
+This abstraction layer makes it possible for the programmers of these systems to write portable programs built with higher level languages.
+The added complexities of using a high level language are small enough compared to the added productivity and convenience for the programmer.
+Some embedded system are also based on an \gls{os}, projects like Raspberry PI\footnote{https://www.raspberrypi.org/} and Tessel\footnote{https://tessel.io/} employs reduced Linux versions to run Python and JavaScript respectively.
+
+In some embedded system, these complexities which leads to lower performance and higher memory usage, makes it hard to benefit from an \gls{os}.
+With the absence of an \gls{os}, applications for embedded system are usually written in a lower level languages.
+These languages must provide the low level control which is needed by the programmer to interact directly with the hardware.
+A well known project running without an \gls{os} is the electronic prototyping platform Arduino\footnote{http://www.arduino.cc/}.
+The EFM32 {\emlib} library used in this thesis is also a platform for {\C} programming without an \gls{os}.
+
+In this thesis we use the words \emph{embedded} and \emph{bare-metal} interchangeably, to refer to execution of code directly on the hardware, without the abstraction of an \gls{os}.
+This is the only execution mode that we have targeted in this project.
+
+\subsection{Programming Model}
+A common programming model used in embedded systems is the \emph{Event-driven} programming model.
+In the this model the program is controlled by events which in turn triggers actions.
+Within an embedded system, these events are hardware interrupt and the actions are handler functions dispatched by the runtime system.
+A typical example of this event-action pair is interrupt given by pressing a button and the program turning on a \gls{LED}.
+Other events that triggers interrupt handlers in an embedded system is, timers, sensors with available data and communication peripherals ready to receive or send data.
+This programming model is successfull in these systems as the peripherals controlled by the \gls{mcu} usually requires time to perform its operation, and the asyncronous nature of the model lets the \gls{cpu} avoid the busy-waits implied by using a more synchronous model.
+
+\subsection{Programming Language}
+
+In embedded systems low level languages have been preferred.
+Traditionally a large portion of code bases consists of C, Assembly, and C++.
+As seen in \autoref{fig:vdc:langs} according to VDC Research \footnote{http://www.oracle.com/us/technologies/java/java-embedded-market-wp-2179018.pdf} the usage of C and assembly are on the way down in favor for higher level languages like C++, Java, and C\#.
```

Comment:
Dette burde også legges til en ekstra linje som sier at denne researchen er for språk for _embedded_ systemer!
Akkurat nå virker det som at det bare er mer populært med høyere-nivå språk generelt, ikke nødvendigvis at denne økningen skjer for embedded